File: src/utils/linkGenerator.ts
/**
 * Extracts AT URI components from various URL formats and generates aturi.to links
 */

import { resolveHandleToDid } from './didResolver';

interface AtUriComponents {
  identifier: string; // DID or handle
  collection?: string;
  rkey?: string;
}

/**
 * Extracts AT URI components from a URL or AT URI string
 * 
 * Universal AT URI pattern detection (works with any domain):
 * - https://anydomain.com/at://did:plc:xxx/collection/rkey
 * - https://anydomain.com/at:/did:plc:xxx/collection/rkey
 * - https://anydomain.com/did:plc:xxx/collection/rkey
 * - https://anydomain.com/handle.bsky.social/collection/rkey
 * 
 * Specific platform formats also supported:
 * - https://bsky.app/profile/did:plc:xxx
 * - https://bsky.app/profile/handle.bsky.social/post/rkey
 * - https://blacksky.community/profile/handle/post/rkey
 * - https://anisota.net/profile/handle/post/rkey
 * - https://anisota.net/explorer/handle/collection/rkey
 * - https://reddwarf.app/profile/handle/post/rkey
 * - https://leaflet.pub/p/identifier
 * - https://margin.at/domain.com/annotation/rkey (maps to at.margin.annotation)
 * - https://margin.at/domain.com/collection/rkey (maps to at.margin.collection)
 * - https://semble.so/profile/identifier
 * - https://witchsky.app/profile/handle/post/rkey
 * - https://catsky.social/profile/handle/post/rkey
 * - https://deer.social/profile/handle/post/rkey
 * - at://did:plc:xxx/app.bsky.feed.post/rkey
 */
export function extractAtUriComponents(input: string): AtUriComponents | null {
  const trimmedInput = input.trim();
  
  // Case 1: Native AT URI format (at://...)
  if (trimmedInput.startsWith('at://')) {
    const withoutProtocol = trimmedInput.substring(5); // Remove "at://"
    const parts = withoutProtocol.split('/');
    
    if (parts.length === 1) {
      // Just a profile: at://did:plc:xxx or at://handle.bsky.social
      return { identifier: parts[0] };
    } else if (parts.length === 3) {
      // Full record: at://identifier/collection/rkey
      return {
        identifier: parts[0],
        collection: parts[1],
        rkey: parts[2],
      };
    }
  }
  
  // Case 2: URL formats (https://...)
  try {
    const url = new URL(trimmedInput);
    const pathname = url.pathname;
    const hostname = url.hostname;
    
    // Universal AT URI pattern detection: /at://identifier/collection/rkey
    if (pathname.startsWith('/at://')) {
      const atUri = pathname.substring(1); // Remove leading "/"
      return extractAtUriComponents(atUri); // Recursive call
    }
    
    // Universal AT URI pattern detection: /at:/identifier/collection/rkey
    if (pathname.startsWith('/at:/')) {
      const parts = pathname.substring(5).split('/'); // Remove "/at:/"
      
      if (parts.length === 1) {
        // Profile only: /at:/identifier
        return { identifier: parts[0] };
      } else if (parts.length === 3) {
        // Full record: /at:/identifier/collection/rkey
        return {
          identifier: parts[0],
          collection: parts[1],
          rkey: parts[2],
        };
      }
    }
    
    // Universal pattern: /did:xxx/collection/rkey or /handle.tld/collection/rkey
    // This catches any domain with a DID or handle-like structure in the path
    const pathParts = pathname.substring(1).split('/').filter(p => p); // Remove leading "/" and split
    
    if (pathParts.length >= 1) {
      const potentialIdentifier = pathParts[0];
      
      // Check if first segment looks like a DID or handle
      const isDid = potentialIdentifier.startsWith('did:');
      const isHandle = !isDid && potentialIdentifier.includes('.') && !potentialIdentifier.includes(' ');
      
      if (isDid || isHandle) {
        if (pathParts.length === 1) {
          // Just identifier: /did:plc:xxx or /handle.bsky.social
          return { identifier: potentialIdentifier };
        } else if (pathParts.length === 3) {
          // Full record: /identifier/collection/rkey
          const collection = pathParts[1];
          const rkey = pathParts[2];
          
          // Validate that collection looks like a lexicon (contains dots)
          if (collection.includes('.')) {
            return {
              identifier: potentialIdentifier,
              collection: collection,
              rkey: rkey,
            };
          }
        }
      }
    }
    
    // Specific domain patterns (for non-AT-URI-like paths)
    
    // Standard /profile/identifier format (bsky.app, blacksky.community, anisota.net, 
    // reddwarf.app, witchsky.app, catsky.social, deer.social)
    if (pathname.startsWith('/profile/')) {
      const parts = pathname.substring(9).split('/'); // Remove "/profile/"
      
      if (parts.length === 1) {
        // Profile only: /profile/identifier
        return { identifier: parts[0] };
      } else if (parts.length === 3 && parts[1] === 'post') {
        // Post: /profile/identifier/post/rkey
        return {
          identifier: parts[0],
          collection: 'app.bsky.feed.post',
          rkey: parts[2],
        };
      } else if (parts.length === 3 && parts[1] === 'lists') {
        // List: /profile/identifier/lists/rkey
        return {
          identifier: parts[0],
          collection: 'app.bsky.graph.list',
          rkey: parts[2],
        };
      }
    }
    
    // Anisota explorer format: /explorer/identifier/collection/rkey
    if (pathname.startsWith('/explorer/') && hostname === 'anisota.net') {
      const parts = pathname.substring(10).split('/'); // Remove "/explorer/"
      
      if (parts.length === 1) {
        // Profile only
        return { identifier: parts[0] };
      } else if (parts.length === 3) {
        // Full record
        return {
          identifier: parts[0],
          collection: parts[1],
          rkey: parts[2],
        };
      }
    }
    
    // Leaflet format: /p/identifier
    if (pathname.startsWith('/p/')) {
      const parts = pathname.substring(3).split('/'); // Remove "/p/"
      
      if (parts.length === 1) {
        return { identifier: parts[0] };
      }
    }
    
    // Margin.at format: /domain/recordType/rkey
    // Maps to at.margin.{recordType} lexicons
    if (hostname === 'margin.at') {
      const parts = pathname.substring(1).split('/').filter(p => p); // Remove leading "/" and split
      
      if (parts.length === 3) {
        const domain = parts[0];
        const recordType = parts[1];
        const rkey = parts[2];
        
        // Valid margin record types: annotation, highlight, bookmark, collection, collectionItem, reply, like
        const validMarginTypes = ['annotation', 'highlight', 'bookmark', 'collection', 'collectionitem', 'reply', 'like'];
        
        if (validMarginTypes.includes(recordType.toLowerCase())) {
          // The domain in the URL is actually the handle
          // Collection is at.margin.{recordType}
          return {
            identifier: domain,
            collection: `at.margin.${recordType.toLowerCase()}`,
            rkey: rkey,
          };
        }
      } else if (parts.length === 2 && parts[0] === 'profile') {
        // Profile format: /profile/identifier
        return { identifier: parts[1] };
      } else if (parts.length === 1) {
        // Just a domain/handle
        return { identifier: parts[0] };
      }
    }
    
    // Semble format: /profile/identifier
    if (hostname === 'semble.so' && pathname.startsWith('/profile/')) {
      const parts = pathname.substring(9).split('/'); // Remove "/profile/"
      
      if (parts.length === 1) {
        return { identifier: parts[0] };
      }
    }
    
    // Legacy /at/ format: /at/identifier or /at/identifier/collection/rkey
    if (pathname.startsWith('/at/')) {
      const parts = pathname.substring(4).split('/'); // Remove "/at/"
      
      if (parts.length === 1) {
        // Profile only
        return { identifier: parts[0] };
      } else if (parts.length === 3) {
        // Full record
        return {
          identifier: parts[0],
          collection: parts[1],
          rkey: parts[2],
        };
      }
    }
    
    // atp.tools format: /record/identifier/collection/rkey or /profile/identifier
    if (pathname.startsWith('/record/')) {
      const parts = pathname.substring(8).split('/'); // Remove "/record/"
      
      if (parts.length === 1) {
        // Just identifier
        return { identifier: parts[0] };
      } else if (parts.length === 3) {
        // Full record
        return {
          identifier: parts[0],
          collection: parts[1],
          rkey: parts[2],
        };
      }
    }
    
  } catch {
    // Not a valid URL, might be a bare identifier
  }
  
  // Case 3: Bare identifier (DID or handle)
  if (trimmedInput.startsWith('did:')) {
    return { identifier: trimmedInput };
  }
  
  // Case 4: Handle-like string (contains dots and no slashes)
  if (trimmedInput.includes('.') && !trimmedInput.includes('/')) {
    return { identifier: trimmedInput };
  }
  
  return null;
}

/**
 * Generates an aturi.to link from AT URI components.
 *
 * Canonical URLs include the `/profile/` prefix to mirror the bsky.app /
 * anisota.net layout. The bare-path forms (`aturi.to/{identifier}/...`) still
 * resolve, but new links should always use `/profile/` so callers like the
 * extension and share UI produce consistent output.
 *
 * @param useAtPrefix - If true, keeps the literal at:// prefix (e.g., aturi.to/at://did:plc:xxx/...)
 */
export function generateAturiLink(components: AtUriComponents, useAtPrefix: boolean = false): string {
  const { identifier, collection, rkey } = components;

  if (useAtPrefix) {
    if (collection && rkey) {
      return `https://aturi.to/at://${identifier}/${collection}/${rkey}`;
    }
    return `https://aturi.to/at://${identifier}`;
  }

  if (collection && rkey) {
    if (collection === 'app.bsky.feed.post') {
      return `https://aturi.to/profile/${identifier}/post/${rkey}`;
    }
    if (collection === 'app.bsky.graph.list') {
      return `https://aturi.to/profile/${identifier}/lists/${rkey}`;
    }
    return `https://aturi.to/profile/${identifier}/${collection}/${rkey}`;
  }

  return `https://aturi.to/profile/${identifier}`;
}

/**
 * Resolves a handle to DID if possible, returns the original if it's already a DID or resolution fails
 */
async function resolveIdentifierToDid(identifier: string): Promise<string> {
  // If it's already a DID, return it
  if (identifier.startsWith('did:')) {
    return identifier;
  }
  
  // Try to resolve handle to DID
  try {
    const did = await resolveHandleToDid(identifier);
    return did || identifier; // Fall back to handle if resolution fails
  } catch (error) {
    console.warn(`Failed to resolve handle ${identifier} to DID:`, error);
    return identifier; // Fall back to handle on error
  }
}

/**
 * Main function to convert any input to an aturi.to link
 * @param useAtPrefix - If true, keeps the literal at:// prefix for full AT URI format
 * @param preferDid - If true, attempts to resolve handles to DIDs (default: false for backwards compatibility)
 */
export async function convertToAturiLink(input: string, useAtPrefix: boolean = false, preferDid: boolean = false): Promise<string | null> {
  const components = extractAtUriComponents(input);
  
  if (!components) {
    return null;
  }
  
  // If preferDid is true, try to resolve handle to DID
  if (preferDid) {
    components.identifier = await resolveIdentifierToDid(components.identifier);
  }
  
  return generateAturiLink(components, useAtPrefix);
}

/**
 * Synchronous version that doesn't resolve handles to DIDs
 * Kept for backwards compatibility
 */
export function convertToAturiLinkSync(input: string, useAtPrefix: boolean = false): string | null {
  const components = extractAtUriComponents(input);
  
  if (!components) {
    return null;
  }
  
  return generateAturiLink(components, useAtPrefix);
}

/**
 * Validates if the input can be converted to an aturi.to link
 */
export function isValidInput(input: string): boolean {
  return extractAtUriComponents(input) !== null;
}

File: ios/Packages/AturiCore/Sources/AturiCore/Waypoints/LinkGenerator.swift
import Foundation

// Port of src/utils/linkGenerator.ts: pull AT URI components out of whatever
// the user pasted or shared (an AT URI, a client URL, a bare handle or DID)
// and build the canonical aturi.to link for them. The async DID-resolving
// variants are not ported; resolution lives in Identity.

/// The identity (and optionally the record) an input names.
public struct AtUriComponents: Hashable, Sendable {
    /// DID or handle.
    public var identifier: String
    public var collection: String?
    public var rkey: String?

    public init(identifier: String, collection: String? = nil, rkey: String? = nil) {
        self.identifier = identifier
        self.collection = collection
        self.rkey = rkey
    }

    public var isRecord: Bool {
        collection != nil && rkey != nil
    }
}

/// The parts of an absolute URL the extractor looks at, parsed by hand.
///
/// JS `new URL('https://x.y/at://did:plc:a/b/c')` keeps the path
/// `/at://did:plc:a/b/c` intact; Foundation's URL rejects or mangles such
/// input, so the extractor splits `scheme://authority/path` itself and only
/// borrows the WHATWG rules that matter for its decisions: the scheme is
/// lowercased, hosts of special schemes are lowercased and get a `/` path
/// when empty, query and fragment are dropped, and an authority whose port is
/// not numeric is not a URL at all (which is why `at://did:plc:x/coll` falls
/// through to the bare-identifier checks exactly as it does in the browser).
private struct LinkGeneratorParsedURL {
    let scheme: String
    let hostname: String
    let pathname: String

    private static let specialSchemes: Set<String> = ["http", "https", "ws", "wss", "ftp", "file"]

    /// Characters JS leaves alone in a path. Everything else (space, quotes,
    /// angle brackets, braces, backtick, controls, non-ASCII) is percent
    /// encoded as UTF-8, and an existing `%` is kept as is.
    private static let pathAllowed: CharacterSet = {
        var set = CharacterSet(charactersIn: UnicodeScalar(0x21)...UnicodeScalar(0x7e))
        set.remove(charactersIn: "\"<>`{}")
        return set
    }()

    init?(_ input: String) {
        guard let separator = input.range(of: "://") else { return nil }
        let rawScheme = input[input.startIndex..<separator.lowerBound]
        guard let first = rawScheme.first, first.isASCII, first.isLetter,
              rawScheme.allSatisfy({ $0.isASCII && ($0.isLetter || $0.isNumber || $0 == "+" || $0 == "-" || $0 == ".") })
        else { return nil }
        let scheme = rawScheme.lowercased()
        let isSpecial = Self.specialSchemes.contains(scheme)

        let rest = input[separator.upperBound...]
        let authorityEnd = rest.firstIndex(where: { $0 == "/" || $0 == "?" || $0 == "#" }) ?? rest.endIndex
        var authority = String(rest[rest.startIndex..<authorityEnd])
        if let at = authority.lastIndex(of: "@") {
            authority = String(authority[authority.index(after: at)...])
        }

        // Split host from port. A bracketed IPv6 literal may carry colons of
        // its own; anything after the closing bracket is the port.
        var host = authority
        var port = ""
        if authority.hasPrefix("[") {
            if let close = authority.firstIndex(of: "]") {
                host = String(authority[authority.startIndex...close])
                let tail = authority[authority.index(after: close)...]
                if tail.hasPrefix(":") { port = String(tail.dropFirst()) } else if !tail.isEmpty { return nil }
            } else {
                return nil
            }
        } else if let colon = authority.firstIndex(of: ":") {
            host = String(authority[authority.startIndex..<colon])
            port = String(authority[authority.index(after: colon)...])
        }
        guard port.allSatisfy({ $0.isASCII && $0.isNumber }) else { return nil }
        if isSpecial && host.isEmpty { return nil }
        if host.contains(" ") { return nil }

        var path = String(rest[authorityEnd...])
        if let stop = path.firstIndex(where: { $0 == "?" || $0 == "#" }) {
            path = String(path[path.startIndex..<stop])
        }
        if isSpecial && path.isEmpty { path = "/" }
        path = path.addingPercentEncoding(withAllowedCharacters: Self.pathAllowed) ?? path

        self.scheme = scheme
        self.hostname = isSpecial ? host.lowercased() : host
        self.pathname = path
    }
}

/// Extracts AT URI components from a URL or AT URI string.
///
/// Universal AT URI pattern detection (works with any domain):
/// - https://anydomain.com/at://did:plc:xxx/collection/rkey
/// - https://anydomain.com/at:/did:plc:xxx/collection/rkey
/// - https://anydomain.com/did:plc:xxx/collection/rkey
/// - https://anydomain.com/handle.bsky.social/collection/rkey
///
/// Specific platform formats also supported:
/// - https://bsky.app/profile/did:plc:xxx
/// - https://bsky.app/profile/handle.bsky.social/post/rkey
/// - https://leaflet.pub/p/identifier
/// - https://margin.at/domain.com/annotation/rkey (maps to at.margin.annotation)
/// - https://semble.so/profile/identifier
/// - at://did:plc:xxx/app.bsky.feed.post/rkey
public func extractAtUriComponents(_ input: String) -> AtUriComponents? {
    let trimmedInput = input.trimmingCharacters(in: .whitespacesAndNewlines)

    // Case 1: native AT URI.
    if trimmedInput.hasPrefix("at://") {
        let parts = trimmedInput.dropFirst(5).split(separator: "/", omittingEmptySubsequences: false).map(String.init)
        if parts.count == 1 {
            return AtUriComponents(identifier: parts[0])
        } else if parts.count == 3 {
            return AtUriComponents(identifier: parts[0], collection: parts[1], rkey: parts[2])
        }
    }

    // Case 2: URL formats.
    if let url = LinkGeneratorParsedURL(trimmedInput) {
        let pathname = url.pathname
        let hostname = url.hostname

        // Universal AT URI pattern: /at://identifier/collection/rkey
        if pathname.hasPrefix("/at://") {
            return extractAtUriComponents(String(pathname.dropFirst()))
        }

        // Universal AT URI pattern, single-slash spelling: /at:/identifier/collection/rkey
        if pathname.hasPrefix("/at:/") {
            let parts = pathname.dropFirst(5).split(separator: "/", omittingEmptySubsequences: false).map(String.init)
            if parts.count == 1 {
                return AtUriComponents(identifier: parts[0])
            } else if parts.count == 3 {
                return AtUriComponents(identifier: parts[0], collection: parts[1], rkey: parts[2])
            }
        }

        // Universal pattern: /did:xxx/collection/rkey or /handle.tld/collection/rkey.
        // Catches any domain with a DID or handle-like structure in the path.
        let pathParts = pathname.dropFirst().split(separator: "/", omittingEmptySubsequences: true).map(String.init)

        if let potentialIdentifier = pathParts.first {
            let isDid = potentialIdentifier.hasPrefix("did:")
            let isHandle = !isDid && potentialIdentifier.contains(".") && !potentialIdentifier.contains(" ")

            if isDid || isHandle {
                if pathParts.count == 1 {
                    return AtUriComponents(identifier: potentialIdentifier)
                } else if pathParts.count == 3 {
                    let collection = pathParts[1]
                    let rkey = pathParts[2]
                    // The collection has to look like a lexicon NSID.
                    if collection.contains(".") {
                        return AtUriComponents(identifier: potentialIdentifier, collection: collection, rkey: rkey)
                    }
                }
            }
        }

        // Specific domain patterns for non-AT-URI-like paths.

        // Standard /profile/identifier format (bsky.app, blacksky.community,
        // anisota.net, reddwarf.app, witchsky.app, deer.social).
        if pathname.hasPrefix("/profile/") {
            let parts = pathname.dropFirst(9).split(separator: "/", omittingEmptySubsequences: false).map(String.init)
            if parts.count == 1 {
                return AtUriComponents(identifier: parts[0])
            } else if parts.count == 3 && parts[1] == "post" {
                return AtUriComponents(identifier: parts[0], collection: "app.bsky.feed.post", rkey: parts[2])
            } else if parts.count == 3 && parts[1] == "lists" {
                return AtUriComponents(identifier: parts[0], collection: "app.bsky.graph.list", rkey: parts[2])
            }
        }

        // Leaflet: /p/identifier
        if pathname.hasPrefix("/p/") {
            let parts = pathname.dropFirst(3).split(separator: "/", omittingEmptySubsequences: false).map(String.init)
            if parts.count == 1 {
                return AtUriComponents(identifier: parts[0])
            }
        }

        // margin.at: /domain/recordType/rkey, mapped to at.margin.{recordType}.
        if hostname == "margin.at" {
            let parts = pathname.dropFirst().split(separator: "/", omittingEmptySubsequences: true).map(String.init)
            if parts.count == 3 {
                let domain = parts[0]
                let recordType = parts[1].lowercased()
                let rkey = parts[2]
                let validMarginTypes: Set<String> = ["annotation", "highlight", "bookmark", "collection", "collectionitem", "reply", "like"]
                if validMarginTypes.contains(recordType) {
                    // The domain in the URL is the handle.
                    return AtUriComponents(identifier: domain, collection: "at.margin.\(recordType)", rkey: rkey)
                }
            } else if parts.count == 2 && parts[0] == "profile" {
                return AtUriComponents(identifier: parts[1])
            } else if parts.count == 1 {
                return AtUriComponents(identifier: parts[0])
            }
        }

        // Semble: /profile/identifier
        if hostname == "semble.so" && pathname.hasPrefix("/profile/") {
            let parts = pathname.dropFirst(9).split(separator: "/", omittingEmptySubsequences: false).map(String.init)
            if parts.count == 1 {
                return AtUriComponents(identifier: parts[0])
            }
        }

        // Legacy /at/ format: /at/identifier or /at/identifier/collection/rkey
        if pathname.hasPrefix("/at/") {
            let parts = pathname.dropFirst(4).split(separator: "/", omittingEmptySubsequences: false).map(String.init)
            if parts.count == 1 {
                return AtUriComponents(identifier: parts[0])
            } else if parts.count == 3 {
                return AtUriComponents(identifier: parts[0], collection: parts[1], rkey: parts[2])
            }
        }

        // atp.tools: /record/identifier/collection/rkey
        if pathname.hasPrefix("/record/") {
            let parts = pathname.dropFirst(8).split(separator: "/", omittingEmptySubsequences: false).map(String.init)
            if parts.count == 1 {
                return AtUriComponents(identifier: parts[0])
            } else if parts.count == 3 {
                return AtUriComponents(identifier: parts[0], collection: parts[1], rkey: parts[2])
            }
        }
    }

    // Case 3: bare DID.
    if trimmedInput.hasPrefix("did:") {
        return AtUriComponents(identifier: trimmedInput)
    }

    // Case 4: handle-like string (contains dots and no slashes).
    if trimmedInput.contains(".") && !trimmedInput.contains("/") {
        return AtUriComponents(identifier: trimmedInput)
    }

    return nil
}

/// Generates an aturi.to link from AT URI components.
///
/// Canonical URLs include the `/profile/` prefix to mirror the bsky.app /
/// anisota.net layout. The bare-path forms (`aturi.to/{identifier}/...`)
/// still resolve, but new links always use `/profile/` so every caller
/// produces consistent output.
///
/// - Parameter useAtPrefix: keep the literal `at://` prefix
///   (`aturi.to/at://did:plc:xxx/...`).
public func generateAturiLink(_ components: AtUriComponents, useAtPrefix: Bool = false) -> String {
    let identifier = components.identifier
    let record: (collection: String, rkey: String)?
    if let collection = components.collection, !collection.isEmpty,
       let rkey = components.rkey, !rkey.isEmpty {
        record = (collection, rkey)
    } else {
        record = nil
    }

    if useAtPrefix {
        if let record {
            return "https://aturi.to/at://\(identifier)/\(record.collection)/\(record.rkey)"
        }
        return "https://aturi.to/at://\(identifier)"
    }

    if let record {
        if record.collection == "app.bsky.feed.post" {
            return "https://aturi.to/profile/\(identifier)/post/\(record.rkey)"
        }
        if record.collection == "app.bsky.graph.list" {
            return "https://aturi.to/profile/\(identifier)/lists/\(record.rkey)"
        }
        return "https://aturi.to/profile/\(identifier)/\(record.collection)/\(record.rkey)"
    }

    return "https://aturi.to/profile/\(identifier)"
}

/// `convertToAturiLinkSync`: any input straight to an aturi.to link, or nil
/// when the input names nothing.
public func convertToAturiLink(_ input: String, useAtPrefix: Bool = false) -> String? {
    guard let components = extractAtUriComponents(input) else { return nil }
    return generateAturiLink(components, useAtPrefix: useAtPrefix)
}

/// Whether the input can be converted to an aturi.to link.
public func isValidInput(_ input: String) -> Bool {
    extractAtUriComponents(input) != nil
}
